SpatialGDK/Source/SpatialGDK/Private/Tests/RPCContainerTest.cpp

```diff
@@ -0,0 +1,232 @@
+	FRPCContainer RPCs;
+
+	RPCs.QueueRPC(MoveTemp(Params1), UnreliableType);
+	RPCs.QueueRPC(MoveTemp(Params2), UnreliableType);
+
+	FProcessRPCDelegate Delegate;
+	Delegate.BindUObject(TargetObject, &UMockObject::ProcessRPC);
+	RPCs.ProcessRPCs(Delegate);
+
+	TestFalse("No queued RPCs of such type", RPCs.ObjectHasRPCsQueuedOfType(ObjecRef.Entity, UnreliableType));
+
+    return true;
+}
+
+TEST(FRPCContainer, GIVEN_a_container_WHEN_two_values_of_different_type_have_been_added_THEN_they_are_queued)
+{
+	UMockObject* TargetObject = NewObject<UMockObject>();
+
+	FPendingRPCParamsPtr ParamsUnreliable = CreateMockParameters(TargetObject, UnreliableType);
+	FUnrealObjectRef ObjecRefUnreliable = ParamsUnreliable->ObjectRef;
+
+	FPendingRPCParamsPtr ParamsReliable = CreateMockParameters(TargetObject, ReliableType);
+	FUnrealObjectRef ObjecRefReliable = ParamsReliable->ObjectRef;
+
+	FRPCContainer RPCs;
+
+	RPCs.QueueRPC(MoveTemp(ParamsUnreliable), UnreliableType);
+	RPCs.QueueRPC(MoveTemp(ParamsReliable), ReliableType);
+
+	TestTrue("Queued RPCs of such type", RPCs.ObjectHasRPCsQueuedOfType(ObjecRefUnreliable.Entity, UnreliableType));
+	TestTrue("Queued RPCs of such type", RPCs.ObjectHasRPCsQueuedOfType(ObjecRefReliable.Entity, ReliableType));
+
+    return true;
+}
+
+TEST(FRPCContainer, GIVEN_a_container_WHEN_two_values_of_different_type_have_been_added_and_processed_THEN_nothing_is_queued)
+{
+	UMockObject* TargetObject = NewObject<UMockObject>();
+	FUnrealObjectRef ObjecRef = GenerateObjectRef(TargetObject);
+	FPendingRPCParamsPtr ParamsUnreliable = CreateMockParameters(TargetObject, UnreliableType);
+	FPendingRPCParamsPtr ParamsReliable = CreateMockParameters(TargetObject, ReliableType);
+
+	FRPCContainer RPCs;
+
+	RPCs.QueueRPC(MoveTemp(ParamsUnreliable), UnreliableType);
+	RPCs.QueueRPC(MoveTemp(ParamsReliable), ReliableType);
+
+	FProcessRPCDelegate Delegate;
+	Delegate.BindUObject(TargetObject, &UMockObject::ProcessRPC);
+	RPCs.ProcessRPCs(Delegate);
+
+	TestFalse("No queued RPCs of such type", RPCs.ObjectHasRPCsQueuedOfType(ObjecRef.Entity, UnreliableType));
+	TestFalse("No queued RPCs of such type", RPCs.ObjectHasRPCsQueuedOfType(ObjecRef.Entity, ReliableType));
+
+    return true;
+}
+
+TEST(FRPCContainer, GIVEN_a_container_WHEN_six_values_of_different_type_have_been_added_and_processed_THEN_they_have_been_processed_in_order)
+{
+	UMockObject* TargetObject = NewObject<UMockObject>();
+	FUnrealObjectRef ObjectRef = GenerateObjectRef(TargetObject);
+	FRPCContainer RPCs;
+
+	TMap<ESchemaComponentType, TArray<uint32>> RPCIndices;
+
+	for (int i = 0; i < 4; ++i)
+	{
+		FPendingRPCParamsPtr ParamsUnreliable = CreateMockParameters(TargetObject, UnreliableType);
+		FPendingRPCParamsPtr ParamsReliable = CreateMockParameters(TargetObject, ReliableType);
+		FPendingRPCParamsPtr ParamsCrossServer = CreateMockParameters(TargetObject, CrossServerType);
+		FPendingRPCParamsPtr ParamsMulticast = CreateMockParameters(TargetObject, MulticastType);
+
```

Comment:
Is it significant that all these different types need to be tested here? The fact you're looping suggests not.. its just the sequence that the RPCs are processed that matters, right? In which case I would be tempted to reduce it to just 2 types and even to hide those behind a couple of variables like AnySchemaComponentType, AnyOtherSchemaComponentType or something like that.

Reply:
I agree, 2 types should be enough.

Reply:
Left 2 types only as suggested.